kernel/objects/rapl-driver-intel/objects/RaplDriverIntel.cc

```diff
@@ -0,0 +1,254 @@
+/* -*- mode:C++; indent-tabs-mode:nil; -*- */
+/* MIT License -- MyThOS: The Many-Threads Operating System
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Copyright 2020 Philipp Gypser and contributors, BTU Cottbus-Senftenberg
+ */
+
+#include "objects/RaplDriverIntel.hh"
+#include "mythos/InvocationBuf.hh"
+#include "util/assert.hh"
+#include "boot/mlog.hh"
+#include "mythos/protocol/RaplDriverIntel.hh"
+#include "mythos/HostInfoTable.hh"
+#include "objects/TypedCap.hh"
+#include "cpu/ctrlregs.hh"
+#include "objects/IntelRegs.hh"
+//#include "objects/IFrame.hh"
+//#include <cmath>
+
+
+namespace mythos {
+
+  extern PhysPtr<HostInfoTable> hostInfoPtrPhys SYMBOL("_host_info_ptr");
+
+  RaplDriverIntel::RaplDriverIntel(){
+    MLOG_INFO(mlog::boot, "RAPL driver startet");
+    // check it is an Intel cpu
+    auto r = x86::cpuid(0);
+    char str[] = "0123456789ab";
+    *reinterpret_cast<uint32_t*>(&str[0]) = r.ebx;
+    *reinterpret_cast<uint32_t*>(&str[4]) = r.edx;
+    *reinterpret_cast<uint32_t*>(&str[8]) = r.ecx;
+    
+    isIntel = true;
+    char intelStr[] = INTEL_STR;
+    for(int i = 0; i < 12; i++){
+      if(str[i] != intelStr[i]){
+        isIntel = false;
+        break;
+      }
+    }
+
+    if(isIntel){
+      MLOG_INFO(mlog::boot, "Congratulations, you are the proud owner of an Intel CPU.");
+    }else{
+      MLOG_ERROR(mlog::boot, "Unsupported processor manufacturer -> ", str);
+    }
+
+    // check cpu family
+    cpu_fam = bits(x86::cpuid(1).eax, 11, 8);
+    MLOG_INFO(mlog::boot, "CPU family =", cpu_fam);
+
+    // check extended cpu model 
+    cpu_model = (bits(x86::cpuid(1).eax, 19, 16) << 4) + bits(x86::cpuid(1).eax, 7, 4);
```

Comment:
would be nice to have these in cpu/cpu/ctrlregs-amd64 to be able to reuse the processor, family, and model queries, and make this here more clear to read.

Reply:
done